pendulum-chain/spacewalk
# 401 fix issues of connecting to stellar overlay network on testnet

closes #401;
Coming from https://github.com/pendulum-chain/spacewalk/pull/409

## General overview of the changes:
This is to support both **_`StellarMessage::TxSet`_** and **_`StellarMessage::GeneralizedTxSet`_** after sending a `StellarMessage::GetTxSet`.

* For pallets that use the `spacewalk-primitives` lib, I **removed** the `substrate-stellar-sdk` dependency.  
   * To be usable across all pallets, the features **_`offchain`_** and **_`all-types`_** of `substrate-stellar-sdk` are used.
* Validating stellar transactions will now require the new **_`TransactionSetType`_**. 
    * https://github.com/pendulum-chain/substrate-stellar-sdk/pull/21 for reference.
* Broke down the `type.rs` inside _vault/oracle_ to multiple files contained in the new `type` dir.
    * The `TxSetMap` change its value type from `TransactionSet` to **_`TransactionSetType`_**
       *  Note: 
          * Affects all unit tests for `TxSetMap`. They have been _commented out_, marked as **_`todo`_**
          * Removed old and inapplicable test files:
              * clients/vault/resources/test/tx_sets
              * clients/vault/resources/test/tx_sets_for_testing
* The changes in oracle are to support the **_`GeneralizedTransactionSet`_** message and to use **_`TransactionSetType`_** when creating proofs.      
* Log errors in `stellar-relay-lib`           
* Updated the metadata-standalone.scale



## How to begin the review:
1. `pallet` dir
   * all affected cargo.tomls **MUST HAVE** `spacewalk-primitives` as a dependency.
     * these cargo.tomls **DO NOT** use `substrate-stellar-sdk`
   * all affected pallets  **MUST REPLACE** `substrate_stellar_sdk` to **_`primitives::stellar`_**
   * all `fn validate_stellar_transaction(...)` **MUST REPLACE** `TransactionSet` datatype param to **_`TransactionSetType`_**
   * all `construct_from_raw_encoded_xdr()`  **MUST REPLACE** `TransactionSet` datatype generic to **_`TransactionSetType`_**
2. `vault/oracle` dir
   * `types` dir
     * `limited_fifo_map.rs` - `TxSetMap` **MUST REPLACE**  `TransactionSet` to **_`TransactionSetType`_**
  * `storage/impls.rs` - comment out all `TxSetMap` or `TxSetFIleHandler` tests
  * `collector` dir
    * `proof_builder.rs` - `Proof` struct  **MUST REPLACE**  `TransactionSet` to **_`TransactionSetType`_**
    * `collector.rs`
      * update `fn add_txset()` to support 2 types: `TransactionSet`, `GeneralizedTransactionSet`
      * the tests will now use new samples from _vault/resources/samples_ , and disable any `TxSetMap` tests ( _`todo`_)
    * `agent.rs` - should handle both `StellarMessage::TxSet` and `StellarMessage::GeneralizedTxSet` messages
3. `stellar-relay-lib`
   * `service.rs`  - functions called inside a `tokio::spawn()` of `overlay_connection.rs` should log the errors instead of propagating them:
      * `fn receiving_service(...)`
      * `fn connection_handler(...)`
   * tests/mod.rs
     * update _`stellar_overlay_should_receive_tx_set`_ test case to use the testnet rather than mainnet (`overlay_infos(false);`) 
        * add test to check the hashes of both `StellarMessage::GetTxSet` and `StellarMessage::GeneralizedTxSet` are THE SAME
